Plutarch/Verify.hs

```diff
@@ -0,0 +1,277 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-unticked-promoted-constructors #-}
+{-# OPTIONS_GHC -Wredundant-constraints #-}
+
+module Plutarch.Verify (
+  PTryFrom (PExcess, ptryFrom),
+  pcheckType,
+  pcheckByteStr,
+  pcheckInt,
+  pcheckList,
+  pcheckMap,
+) where
+
+import Plutarch.Builtin (
+  PAsData,
+  PBuiltinList,
+  PBuiltinMap,
+  PBuiltinPair,
+  PData,
+  PIsData (pfromData),
+  pasByteStr,
+  pasInt,
+  pasList,
+  pasMap,
+  pdata,
+  pforgetData,
+  pfstBuiltin,
+  ppairDataBuiltin,
+  psndBuiltin,
+ )
+
+import Plutarch.ByteString (PByteString)
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal.Other (
+  POpaque,
+  PType,
+  Term,
+  pcon,
+  perror,
+  pforce,
+  phoistAcyclic,
+  plam,
+  plet,
+  (#),
+  type (:-->),
+ )
+import Plutarch.Unit (PUnit (PUnit))
+
+import Plutarch.DataRepr.Internal (
+  PDataRecord,
+  PLabeledType ((:=)),
+  pdcons,
+  pdnil,
+ )
+
+import Plutarch.Bool (pif, (#==))
+
+import Plutarch.List (
+  phead,
+  pmap,
+  ptail,
+ )
+
+import Plutarch.Unsafe (punsafeBuiltin, punsafeCoerce)
+import qualified PlutusCore as PLC
+
+import Plutarch.TermCont (TermCont, tcont, unTermCont)
+
+{- |
+    Note: PAsData POpaque ~ PData
+-}
+
+----------------------- The class PTryFrom ----------------------------------------------
+
+{- |
+    This checks the datastructure for validity.
+    If you don't care about parts of the datastructure
+    don't verify those parts, just let it return a PData
+    instead
+    Be aware this might get really expensive, so only
+    use it if you cannot establish trust otherwise
+    (e.g. via only checking a part of your Data with
+    PTryFrom)
+-}
+class PTryFrom (a :: PType) (b :: PType) where
+  --     "safest" type --^            ^-- target type
+  type PExcess a b :: PType
+  ptryFrom :: Term s a -> TermCont s (Term s b, Term s (PExcess a b))
```

Comment:
I'm not clear what `PExcess` is for, if I'm understanding right, it's so you can choose between checking vs unwrapping the structure of `PData`? Is the first item returned always `punsafeCoerce`? 
Something feels a bit awkward about this. 
I think it might make more sense to make the `PExcess` type more primary (as the second param of the typeclass)

Reply:
A data type rather than a tuple might make things clearer, too. Something like:
```haskell
data CheckedAs a b s = CheckedAs { outer :: Term s a, inner :: Term s b } 
```
(naming is hard, but I think `outer` / `inner` is pretty good)

Reply:
See my comment about the law. It's not about outer vs inner, it's about extra information that the checking process finds, but isn't actually relevant for the law.

Reply:
The semantics differ per instance.

Reply:
@t1lde The main idea about it is that I prove that 
`PData ~ PAsData PSomething` holds, in the process I will, in many cases, unwrap the `PAsData PSomething`, to safe work, I use the associated type to safe the actual work done and then just `punsafeCoerce` the input instead of rewrapping it, because if we have the second element of the pair without crashes we can actually be sure that the verification succeeded.

Reply:
We're not proving that `PData ~ PAsData PSomething` holds, though I assume you're writing that as a lazy notation.
To make it entirely clear and to avoid confusion, we're proving that for an `x` where `x \in PData`, `x \in PAsData something`.